2751 Integrate database lookup tables

carlosfelix2 opened the pull request:
This PR integrates the database lookup tables to be handled just like file-based tables.  The result is that tables can be either file-based or databased backed.  File-based take precedence over database.  To reload the database lookup tables, the use of the metadata singleton is used to trigger a check on that last time the database was checked, and when that time expires then it looks to the database to check for any updates.

Test Steps:
We are going to be using the LIVD table to test that the database tables can be used instead of file-based tables as well as loading updates.
1. Start a terminal and go to the `prime-router` folder.
2. Run your local RS instance.
3. Send the `./src/testIntegration/resources/datatests/CSV_to_HL7/sample-single-pdi-20210608-0002.csv` test file to the reports endpoint using the `simple_report` client.
4. Wait until the file is fully processed and verify that the pima-az-covid file sent to the `build/sftp` folder has an Ordered_test_name of `SARS-CoV-2 (COVID-19) Ag [Presence] in Respiratory specimen by Rapid immunoassay`.  This is the baseline we will compare to for the following tests.
5. Create a new lookup table by running the following command: 
`./prime lookuptables create -n LIVD-SARS-CoV-2-2021-09-29 -i ./metadata/tables/LIVD-SARS-CoV-2-2021-09-29.csv`
and answering Y at the prompt.
6. Shutdown your RS instance.
7. Move the LIVD table file we used to the prime-router folder by running 
`mv ./metadata/tables/LIVD-SARS-CoV-2-2021-09-29.csv ./`
This will remove that table's file from the system and allow RS to load the table from the database.
8. Run `gradle clean package` and then start your local RS instance.
9. Send the `./src/testIntegration/resources/datatests/CSV_to_HL7/sample-single-pdi-20210608-0002.csv` test file again to the reports endpoint using the `simple_report` client.
10. Wait until the file is fully processed and verify that the latest pima-az-covid file sent to the `build/sftp` folder has an Ordered_test_name of `SARS-CoV-2 (COVID-19) Ag [Presence] in Respiratory specimen by Rapid immunoassay`. 
11. Edit the file `./LIVD-SARS-CoV-2-2021-09-29.csv`, find the record for `BinaxNOW COVID-19 Ag Card` and change the column `Test Ordered LOINC Long Name` to a different text like `MY TEST SARS-CoV-2 (COVID-19) Ag [Presence] in Respiratory specimen by Rapid immunoassay`. 
12. Create a new lookup table by running the following command: `./prime lookuptables create -n LIVD-SARS-CoV-2-2021-09-29 -i ./LIVD-SARS-CoV-2-2021-09-29.csv` then answer Y at the prompt.
13. Activate the new version by running `./prime lookuptables activate -n LIVD-SARS-CoV-2-2021-09-29 -v 2` (replace the number 2 in `-v 2` with the proper new version if not 2). 
14. Wait 40 seconds to allow for the poll timer to expire, so the new table is loaded on the next run.
15. Send the `./src/testIntegration/resources/datatests/CSV_to_HL7/sample-single-pdi-20210608-0002.csv` test file to the reports endpoint using the `simple_report` client.
16. Wait until the file is fully processed and verify that the latest pima-az-covid file sent to the `build/sftp` folder has an Ordered_test_name of `MY TEST SARS-CoV-2 (COVID-19) Ag [Presence] in Respiratory specimen by Rapid immunoassay`.
17. Revert your repo back to normal by running the following commands:
`rm ./LIVD-SARS-CoV-2-2021-09-29.csv`
`git checkout origin/master ./metadata/tables/LIVD-SARS-CoV-2-2021-09-29.csv`

## Changes
- Integrate database lookup tables - this is done loading the data as Lookup Table objects, which are then handled the same way as the file-based tables.
- Moved the table classes into a metadata folder, aiming to move more in the future to organize the code a little bit more
- Made Metadata a singleton - previously Metadata was being called everywhere by its constructor that loaded everything every time.  Tried to make as much as possible private, so this does not happen again.
- Check for table updates within the Metadata getinstance method.  We are unable to have an outside trigger reload the tables as we have multiple threads and instances running.  Every function that needs metadata will have to invoke the getInstance() function which will trigger a table check at an interval.
 - - Blank out tables that are no longer in the DB or active as removing from memory may break things
 - - Add any newly activated tables to open to a future with some dynamic changes to other portions
- Some other minor cleanups

## Checklist

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [x] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [x] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Fixes
- #2751

## To Be Done
- #2885 -  need to load test data into the DB for local use

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

jimduff-usds commented on prime-router/src/main/kotlin/Metadata.kt:
```diff
@@ -357,23 +439,32 @@ class Metadata {
+        /**
+         * The amount of seconds to wait before tables are checked again for updates.
+         */
+        private val tablePollInternalSecs = if (Environment.get() == Environment.PROD) 300 else 30
```
I've had plans for months to expand the settings table or add a new table to enable per-environment settings such as these.  (Or, it occurs to me that we could use the LookupTable to store per-environment variables.  ;) )

carlosfelix2 replied:
We can do that, just need to plan it good.  Some of the values will not change, though.

carlosfelix2 replied:
I like this.  I will need to find out how we tell DevOps how to set that up

jimduff-usds reviewed (changes requested):
I got a couple unit test failures on 
`gradlew clean package`

```
> Task :test

gov.cdc.prime.router.MetadataTests load database lookup table updates test() FAILED (2.6s)

  io.mockk.MockKException: no answer found for: DatabaseLookupTableAccess(#1).fetchTableList(false)
      at app//gov.cdc.prime.router.MetadataTests.load database lookup table updates test(MetadataTests.kt:295)

gov.cdc.prime.router.MetadataTests check for database lookup table updates test() FAILED

  io.mockk.MockKException: no answer found for: DatabaseLookupTableAccess(#2).fetchTableList(false)
      at app//gov.cdc.prime.router.MetadataTests.check for database lookup table updates test(MetadataTests.kt:279)


FAILURE: Executed 332 tests in 1m 28s (2 failed, 3 skipped)
```

Here's a more illuminating stack trace, from when I run one of the tests inside IntelliJ:

```
no answer found for: DatabaseLookupTableAccess(#1).fetchTableList(false)
io.mockk.MockKException: no answer found for: DatabaseLookupTableAccess(#1).fetchTableList(false)
	at app//io.mockk.impl.stub.MockKStub.defaultAnswer(MockKStub.kt:93)
	at app//io.mockk.impl.stub.MockKStub.answer(MockKStub.kt:42)
	at app//io.mockk.impl.recording.states.AnsweringState.call(AnsweringState.kt:16)
	at app//io.mockk.impl.recording.CommonCallRecorder.call(CommonCallRecorder.kt:53)
	at app//io.mockk.impl.stub.MockKStub.handleInvocation(MockKStub.kt:266)
	at app//io.mockk.impl.instantiation.JvmMockFactoryHelper$mockHandler$1.invocation(JvmMockFactoryHelper.kt:23)
	at app//io.mockk.proxy.jvm.advice.Interceptor.call(Interceptor.kt:21)
	at app//gov.cdc.prime.router.azure.DatabaseLookupTableAccess.fetchTableList(DatabaseLookupTableAccess.kt:68)
	at app//gov.cdc.prime.router.azure.DatabaseLookupTableAccess.fetchTableList$default(DatabaseLookupTableAccess.kt:59)
	at app//gov.cdc.prime.router.Metadata.loadDatabaseLookupTables$prime_router(Metadata.kt:361)
	at app//gov.cdc.prime.router.Metadata.<init>(Metadata.kt:85)
	at app//gov.cdc.prime.router.Metadata.<init>(Metadata.kt:70)
	at app//gov.cdc.prime.router.MetadataTests.load database lookup table updates test(MetadataTests.kt:295)

```

carlosfelix2 commented:
> I got a couple unit test failures on `gradlew clean package`
> 
> ```
> > Task :test
> 
> gov.cdc.prime.router.MetadataTests load database lookup table updates test() FAILED (2.6s)
> 
>   io.mockk.MockKException: no answer found for: DatabaseLookupTableAccess(#1).fetchTableList(false)
>       at app//gov.cdc.prime.router.MetadataTests.load database lookup table updates test(MetadataTests.kt:295)
> 
> gov.cdc.prime.router.MetadataTests check for database lookup table updates test() FAILED
> 
>   io.mockk.MockKException: no answer found for: DatabaseLookupTableAccess(#2).fetchTableList(false)
>       at app//gov.cdc.prime.router.MetadataTests.check for database lookup table updates test(MetadataTests.kt:279)
> 
> 
> FAILURE: Executed 332 tests in 1m 28s (2 failed, 3 skipped)
> ```
> 
> Here's a more illuminating stack trace, from when I run one of the tests inside IntelliJ:
> 
> ```
> no answer found for: DatabaseLookupTableAccess(#1).fetchTableList(false)
> io.mockk.MockKException: no answer found for: DatabaseLookupTableAccess(#1).fetchTableList(false)
> 	at app//io.mockk.impl.stub.MockKStub.defaultAnswer(MockKStub.kt:93)
> 	at app//io.mockk.impl.stub.MockKStub.answer(MockKStub.kt:42)
> 	at app//io.mockk.impl.recording.states.AnsweringState.call(AnsweringState.kt:16)
> 	at app//io.mockk.impl.recording.CommonCallRecorder.call(CommonCallRecorder.kt:53)
> 	at app//io.mockk.impl.stub.MockKStub.handleInvocation(MockKStub.kt:266)
> 	at app//io.mockk.impl.instantiation.JvmMockFactoryHelper$mockHandler$1.invocation(JvmMockFactoryHelper.kt:23)
> 	at app//io.mockk.proxy.jvm.advice.Interceptor.call(Interceptor.kt:21)
> 	at app//gov.cdc.prime.router.azure.DatabaseLookupTableAccess.fetchTableList(DatabaseLookupTableAccess.kt:68)
> 	at app//gov.cdc.prime.router.azure.DatabaseLookupTableAccess.fetchTableList$default(DatabaseLookupTableAccess.kt:59)
> 	at app//gov.cdc.prime.router.Metadata.loadDatabaseLookupTables$prime_router(Metadata.kt:361)
> 	at app//gov.cdc.prime.router.Metadata.<init>(Metadata.kt:85)
> 	at app//gov.cdc.prime.router.Metadata.<init>(Metadata.kt:70)
> 	at app//gov.cdc.prime.router.MetadataTests.load database lookup table updates test(MetadataTests.kt:295)
> ```

I changed the initialization of metadata and forgot to run the unit tests.  Fixed.  Thx

carlosfelix2 commented on prime-router/src/main/kotlin/Metadata.kt:
```diff
@@ -357,23 +439,32 @@ class Metadata {
+
+        /**
+         * Get the singleton instance.
+         */
```
Add @return

carlosfelix2 replied:
Fixed

carlosfelix2 commented on prime-router/src/main/kotlin/Metadata.kt:
```diff
@@ -357,23 +439,32 @@ class Metadata {
+        fun getInstance(): Metadata {
+            // Load any updates to the database lookup tables.
+            // Due to the use of this function, this will get checked at least on every Azure function call
+            singletonInstance.checkForDatabaseLookupTableUpdates()
```
Explore changing this to run at the start at the function.

carlosfelix2 commented on prime-router/src/main/kotlin/metadata/DatabaseLookupTable.kt:
```diff
@@ -0,0 +1,55 @@
+
+    /**
+     * Load the table [version] from the database.
+     */
```
Add @return

carlosfelix2 replied:
Fixed.

clediggins-usds commented on prime-router/src/main/kotlin/Metadata.kt:
```diff
@@ -51,28 +58,48 @@ class Metadata {
 
     /**
      * Useful for test versions of the metadata catalog
+     * @param tableDbAccess database lookup table access for dependency injection purposes
```
Would you mind adding an explicit comment that this is for testing only, and should never be called from any non-test code?

carlosfelix2 replied:
Good idea.  Done.

whytheplatypus commented on prime-router/src/main/kotlin/Metadata.kt:
```diff
@@ -306,6 +337,57 @@ class Metadata {
+                error("Database lookup table list returned an inactive table.")
+
+            // Process existing tables
+            databaseTables.forEach { dbTable ->
```
This is hard to follow as it's written.

maybe  the loop can be flipped?:

```
activeTables.forEach { table ->
  val existingTable = databaseTables[table.name]
  // if it's not in the list it's a new table.
  // check it's version.
```
then handle the cases where an existing table was removed separately?

carlosfelix2 replied:
Maybe this when statement can help

whytheplatypus commented on prime-router/src/main/kotlin/metadata/DatabaseLookupTable.kt:
```diff
@@ -0,0 +1,56 @@
+class DatabaseLookupTable(
+    val name: String,
+    private val tableDbAccess: DatabaseLookupTableAccess = DatabaseLookupTableAccess()
+) : LookupTable(emptyList()), Logging {
```
Rather than inherit from LookupTable what's the common interface they could both implement?

carlosfelix2 replied:
I am not implementing new table operations, plus the LookTable does not have a load.  Even if I create an interface class I would still need to inherit from LookupTable.

carlosfelix2 commented on prime-router/src/main/kotlin/Metadata.kt:
```diff
@@ -306,6 +337,57 @@ class Metadata {
+                    dbTable.loadTable(tableInfo.tableVersion)
+                else if (tableInfo == null) {
+                    // This table is no longer active or was removed.
+                    dbTable.setTableData(emptyList())
```
We cannot remove the table as that will break the schema due to the way the schemas are handled by the code.

whytheplatypus replied:
got it.. so what happens when it's an empty list? I guess could you add some comment here on that gotcha, and what high level changes would have to happen for it not to be an issue anymore?

carlosfelix2 replied:
Good idea.  I find the biggest problem is separation of concern and I think we will have to incrementally get that better to then make fundamental changes to the code.  Not easy!
